### PR TITLE
Add full text search

### PR DESCRIPTION
**Pull request by Marigold**
* add full-text search index
* endpoints for search & sql
* more output formats (csv, json, feather)

### CODE REVIEW

File: README.md [hunk 1]
```diff
@@ -32,10 +32,13 @@ Docs are available at http://127.0.0.1:8000/v1/docs.
 
 ### Sample Queries
 
+Sample queries written in [httpie](https://httpie.io/)
+
 - http GET http://127.0.0.1:8000/health
 - http GET http://127.0.0.1:8000/v1/variableById/data/42539
 - http GET http://127.0.0.1:8000/v1/variableById/metadata/42539
```
Comment (wcox13) [inline README.md:39]: not related to this PR, but would be good to have examples for the `v1/dataset` endpoints too

File: app/v1/data.py [hunk 2]
```diff
@@ -17,9 +17,82 @@
...
+    con, sql: str, type: DATA_TYPES, parameters: list[Any] = []
+) -> Any:
+    # read data in feather format and return it directly in response
+    # NOTE: should be the fastest in theory, but is really slow for unknown reasons
+    if type == "feather_direct":
+        bytes_io = _read_sql_bytes(con, sql, parameters=parameters)
+        # NOTE: this approach is much faster than `StreamingResponse(bytes_io, ...`, how is it possible?
+        # response = StreamingResponse(
+        #     iter([bytes_io.getvalue()]), media_type="application/octet-stream"
+        # )
```
Comment (wcox13) [inline app/v1/data.py]: i'm confused by this comment - is `this approach` the one in the comments or the one in the code?

File: app/v1/data.py [hunk 3]
```diff
@@ -17,9 +17,82 @@
...
+        response = StreamingResponse(bytes_io, media_type="application/octet-stream")
+        response.headers["Content-Disposition"] = "attachment; filename=test.feather"
```
Comment (wcox13) [inline app/v1/data.py]: filename is still `test.feather` (which is my doing haha)

File: app/v1/data.py [hunk 4]
```diff
@@ -17,9 +17,82 @@
...
+    # read data into dataframe and then convert to feather
+    elif type == "feather":
+        bytes_io = io.BytesIO()
+        df = con.execute(sql, parameters=parameters).fetch_df()
+        write_feather(df, bytes_io)
+
+        response = StreamingResponse(
+            iter([bytes_io.getvalue()]), media_type="application/octet-stream"
+        )
+        response.headers["Content-Disposition"] = "attachment; filename=test.feather"
```
Comment (wcox13) [inline app/v1/data.py]: should we have a subroutine for this `StreamingResponse` stuff? looks about the same here and above

File: app/v1/data.py [hunk 5]
```diff
@@ -17,9 +17,82 @@
...
+        # convert to numpy arrays on which ormjson is super fast
+        return df.to_dict(orient="list")
+
+    else:
+        raise HTTPException(status_code=400, detail=f"unknown type {type}")
+
+
+@router.post("/sql")
+def sql_query(sql: str, type: DATA_TYPES = "csv"):
+    """Run arbitrary query on top of our database."""
```
Comment (wcox13) [inline app/v1/data.py:91]: what is the reason for having this endpoint? isn't it easier to just connect to the db to run your sql query?

File: app/v1/search.py [hunk 6]
```diff
@@ -0,0 +1,65 @@
...
+    dataset = "meta_datasets"
+
+
+@router.get("/search")
+def search(term: str, type: SearchType = SearchType.variable, limit: int = 10):
+    con = utils.get_readonly_connection(threading.get_ident())
+
+    # TODO: implement search on other tables too? not sure whether we'll need it yet
+    if type != SearchType.variable:
+        raise NotImplementedError()
```
Comment (wcox13) [inline app/v1/search.py]: nit: add a message? `NotImplementedError("Invalid search type {type}, only searching variables is currently supported")`

**Review by wcox13 — APPROVED** [review]

File: app/v1/data.py (hunk 4 above)
Comment (wcox13) [reply app/v1/data.py]: also another `test.feather` here

File: README.md (hunk 1 above)
Comment (Marigold) [reply README.md:39]: Good point, added covid dataset and maddison metadata.

File: app/v1/data.py (hunk 2 above)
Comment (Marigold) [reply app/v1/data.py]: Oh thanks, clarified

File: app/v1/data.py (hunk 3 above)
Comment (Marigold) [reply app/v1/data.py]: Changed it to `owid.feather`, not much better though...

File: app/v1/data.py (hunk 4 above)
Comment (Marigold) [reply app/v1/data.py]: DRYed, thanks

File: app/v1/data.py (hunk 5 above)
Comment (Marigold) [reply app/v1/data.py:91]: > isn't it easier to just connect to the db to run your sql query?

Maybe? It's not crystal clear how to connect to a remote DuckDB from their docs. You might also need a driver or at least DuckDB installed. Calling API is simpler and lets us have more control (like putting it behind cloudflare, setting timeouts, blocking requests, etc.). Happy to hear your thoughts!

**Pull request merged**